Policies: Adding plugin hook for auto-approval flow #6203

dynamic-entropy opened the pull request:
### Refactored auto-approval logic and introduced algorithm-based approval options
Fix #6203

The code changes in this pull request refactor the auto-approval logic for replication rules.
This introduces a new pluggable algorithms-based approach while retaining the old flow for approval, which was based on `size` and `length` limits of the DIDs (referred to as 'legacy'* in the code)

The changes include:
    - Allowing multi-RSE approval (or not) based on the configured algorithm 
    - The "legacy" auto-approval method is evaluated if no algorithm is specified and if not approved already.

A sample function to auto_approve will have the following signature:
```
def multi_rse_approval(did, rule_attributes, session):
    return True
 ```
 
They would need to be specified in the `get_algorithms` function as part of the Policy package. 
For e.g. `MyPolicyPackage/__init__.py`:
```
def get_algorithms():
    return {
        'lfn2pfn': {
            'cmstfc': cmstfc
        },
        'auto_approve': {
            'global': multi_rse_approval,
            'single': single_rse_approval
        }
    }
```
*Needed to name it to facilitate logging the name of the algorithm that approved the rule.

dynamic-entropy commented:
Hi @bari12 @jamesp-epcc 
Can you please find some time for a review of this? 

Also, do we know if any other community is already using the old auto_approve flow? There is a tiny bug in the code for it 
https://github.com/rucio/rucio/blob/15d07913a20d360342894f9e2e31d72ad349a640/lib/rucio/core/rule.py#L265
where we check the `bytes` limit against the `length` of did.
Either we can fix this or If no one is using it. Do you think it is a good idea to deprecate auto-approve bytes? 
Since this can be done in plugin scripts?

Cheers
Rahul

bari12 commented on lib/rucio/core/rule.py:
```diff
@@ -254,28 +254,69 @@ def add_rule(dids, account, copies, rse_expression, grouping, weight, lifetime,
-                    elif did.bytes < int(rse_attr.get('auto_approve_bytes', -1)):
-                        auto_approve = True
-                    elif did.length < int(rse_attr.get('auto_approve_files', -1)):
+                configured_algorithm, auto_approve = None, False
```
I think this setup part should really be on root of the module. Otherwise you are reloading the modules every time? @jamesp-epcc I think this is also how it is done in other algorithms?

dynamic-entropy replied:
Hi @bari12 
It will not cause any more reloads than when defined in the root of the module. 
The module will be loaded in the `sys.modules` it will remain there till the module's lifetime.

bari12 replied:
Yes, that's true, but you are still calling the lines above (register_policy_package_algorithms() etc. over and over again. While in comparison, if you look at common.utils with the surl algorithms, there this is only loaded on module load, and then just re-used.

dynamic-entropy replied:
Thank you for pointing this out and clarifying.

bari12 commented on lib/rucio/core/rule.py:
```diff
@@ -254,28 +254,69 @@ def add_rule(dids, account, copies, rse_expression, grouping, weight, lifetime,
+                        'ignore_availability': ignore_availability,
+                    }
+
+                    if auto_approve_algorithm_fn(did, rule_attributes, session):
```
So here you want the algorithm to have a boolean return and if not fall back to the default?

dynamic-entropy replied:
Hi @bari12 I did not understand the question entirely. 
Yes, it's a boolean return. 
If not, it is `False` by default.

`configured_algorithm, auto_approve = None, False`

bari12 replied:
I wonder if we need that? I mean this is basically so you have a fallback to the default algorithm, but I would almost be stricter on this. If the policy one doesnt work, then there just is no auto-approval.

dynamic-entropy replied:
Ok, I can make it stricter

``` 
if configured_algorithm == None
# use default
```

bari12 reviewed (commented):
Added some comments. 
Yes, please fix the one wrong check. Good catch :-D

dynamic-entropy commented:
> Yes, please fix the one wrong check. Good catch :-D

Done

dynamic-entropy commented:
Hi @bari12 

I have addressed the requested changes. Kindly have a look. 
If all is good, I will squash and rebase.

ericvaandering commented:
Is the target for getting this in 1.31 or 32?

dynamic-entropy committed:
Policies: Add BaseClass for policy package algorithms #6203

dynamic-entropy committed:
Policies: Add auto approve plugin; Fix #6203

dynamic-entropy committed:
Housekeeping: Type annotations and pylint errors

Adding type annotations for special variables
These include
- return type for __init__
- type for cls in classmethods

Fixing pylint errors
- raise missing from exception: https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/raise-missing-from.html
- change NoResultFound import from sqlalchemy.orm.exc to sqlalchemy.exc

bari12 commented on lib/rucio/core/plugins.py:
```diff
@@ -0,0 +1,162 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
```
Please remove the duplicated header part. Once is sufficient :-)

dynamic-entropy replied:
Done

bari12 reviewed (commented):
Only one minor comment, otherwise gtg! Thanks!

dynamic-entropy commented:
Hi @bari12 
Sorry for the prolonged delay. 
I have made the following changes:
- Remove the extra headers from the plugins.py file. 
- rebased the branch onto current master
- Added a new commit with some basic housekeeping based on Pylint complains
   - Missing "from" when exception chaining: https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/raise-missing-from.html
   - Changed import source for `NoResultFound` as per https://pydoc.dev/sqlalchemy/latest/sqlalchemy.exc.NoResultFound.html

jamesp-epcc commented:
Apologies for coming very late to this. I somehow missed the notification for being tagged in it and only found it by accident much later. I can't really comment on the auto-approval side of things as I'm not familiar with that, but from the policy package side this looks good. The class for loading algorithms from a policy package looks like an improvement on how this was done before.
